L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  virtual void NotifyWhenHasError(std::function<void()> upcall) = 0;
+
+  enum NotificationOutcome {
+    NOTIFY_FD_BECAME_READY,
+    NOTIFY_TIMED_OUT
+  };
+
+  // Similar to NotifyWhen*, but with a deadline. If the file descriptor state 
+  // changes before the deadline, then upcall will be called with 
+  // NOTIFY_FD_BECAME_READY. If the deadline expires before the descriptor state 
+  // changes, then upcall will be called with NOTIFY_TIMED_OUT.
+  virtual void NotifyWhenReadableOrTimeout(std::function<void(NotificationOutcome)> upcall, gpr_timespec deadline) = 0;
+  virtual void NotifyWhenWritableOrTimeout(std::function<void(NotificationOutcome)> upcall, gpr_timespec deadline) = 0;
+
+  // Clears the descriptor's status. This allows the NotifyWhen* calls to 
+  // distinguish between old and new events.
+  virtual void ClearReadable() = 0;
+  virtual void ClearWritable() = 0;
+  virtual void ClearHasError() = 0;
+
+  // Forces the descriptor’s status to change, and runs the closures if registered. 
+  // This is especially useful for tearing down closures associated with this 
+  // descriptor by forcing them to run without waiting for the event on the 
+  // underlying socket.
+  virtual void ForceReadable() = 0;
+  virtual void ForceWritable() = 0;
+  virtual void ForceHasError() = 0;
+
+  // Returns a human-readable description of the file descriptor state.
+  virtual string ToString() const = 0;
+};
+
+class EpollEventManagerInterface: public BaseEventManagerInterface {
+ public:
+  virtual ~EpollEventManagerInterface();
+
+  virtual EpollDescriptorInterface* RegisterFileDescriptor(int fd) = 0;
+
+  virtual void DeleteDescriptor(EpollDescriptorInterface* d) = 0;
+
+  // Adds a new one-shot alarm with a slack, which will be scheduled at the given 
+  // deadline plus the timespan specified in slack. Reducing the precision of 
+  // alarms through a timer slack allows the EventManager to perform 
+  // timer-coalescing, which minimizes thread wake-up and has the potential of 
+  // substantially reducing CPU utilization.
+  virtual void ScheduleAlarmInWithSlack(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, TaskKey* key, gpr_timespec slack) = 0;
+};
+```
+
+Note that there will be no native implementation provided by gRPC for the epoll based interface in the first step. Instead, gRPC will provide a default event manager implementation based on libuv. The libuv EM implementation will be applied to all the gRPC languages based on gRPC core, and the users may choose to override it with their custom implementation of the EM interface. gRPC may release more native event manager implementations down the road map.
+```cpp
+// Represents a file descriptor when using libuv.
+class LibuvDescriptorInterface {
+ public:
+  virtual ~LibuvDescriptorInterface();
+
+  // Returns the underlying TCP or UDP handle. Valid until Close() is called
+  // (i.e. not thread-safe with Close()).
+  virtual uv_handle_t fd() const = 0;
```

Comment:
File descriptors are used directly in Core's code. E.g., [local credentials](https://github.com/grpc/grpc/blob/master/src/core/lib/security/security_connector/local/local_security_connector.cc#L73). If we return anything else than system fd, some features might be impacted.

Reply:
I am not familiar with local credentials, but according to @guantaol local credentials need to move away from directly using fd's. @markdroth this might be a blocking issue for EM?

Reply:
I think we need to do one of two things here:

1. Provide a common inteface for descriptors that exposes source and peer addresses, and expose that from the endpoint.

2. Change the endpoint API to stop exposing the underlying descriptor and instead expose methods to return the source and peer addresses.

My inclination would be to prefer option 1, since we are probably going to need to have a common descriptor interface anyway, as per discussion below.  And option 2 would still require doing that work anyway.

Note that, as mentioned below, the current `grpc_endpoint` API is part of iomgr and will eventually be going away.  Ideally, it should be replaced with a C++ API that is better designed to work in the EM world.  But it might be that the best transition path is to make changes to the existing API and then convert it to C++ later.  We can figure out those details later, though.